src/lib/scheduler/abstract_task.cpp

```diff
@@ -144,4 +147,43 @@ void AbstractTask::_on_predecessor_done() {
   }
 }
 
+bool AbstractTask::_try_transition_to(TaskState new_state) {
+  // Lock critical section
+  auto lock = std::unique_lock<std::mutex>(_transition_to_mutex);
```

Comment:
Why do we need a lock here, why can't we use the return value of `exchange`?

Reply:
Because we may abort the state transition, hence `_try_transition_to`.

To ensure `_state` stays unchanged between the ifs and `.exchange` we need to lock this function. I could change it to `std::lock_guard` though.

Reply:
Ah, got it. I think some examples would help here.